### ISSUE TITLE
Add 3DS Auth support for Cards

### ISSUE THREAD

**Issue by Fifciu** [develodesign/vsf-payment-stripe#50]
Add 3DS Auth support for Cards. I've already implemented it In some project like a year ago. Here you could find my commits: https://github.com/new-fantastic/vsf-payment-stripe/commits/master

Look from 25. Feb 2020

**Comment by dimasch** [comment]
@Fifciu Cool feature, thanks! 
Can you create a pull request and i will create a next tag?

**Comment by Fifciu** [comment]
It will be hard for me in the current period. That's why I put a link to commits.

**Comment by dimasch** [comment]
@Fifciu Okay, i will do it by myself.
Thanks again!

**Comment by antonioglass** [comment]
Hey, @dimasch! I installed this module (https://github.com/new-fantastic/vsf-payment-stripe/commits/master) but it cannot get `clientSecret` and there is `stripeError`. Have you tried this module on the latest VSF1 versions?

I also tried installing an original module but nothing happens after I place the order.

I would appreciate if you could help or update this module. Thank you!

**Comment by dimasch** [comment]
@antonioglass 
> Have you tried this module on the latest VSF1 versions?

Yes, sure, but without the 3DS feature. You can check and create a Merge Request with support this feature, welcome!

**Comment by antonioglass** [comment]
@dimasch Thanks for quick reply! Have you faced a similar issue like nothing happens after placing the order? There are no any errors...

**Comment by dimasch** [comment]
@antonioglass 
No, there are things that do not depend on the settings of the frontend and this module, correct configuration is still required the Magento2 module for Stripe and the api related settings (integration section)

**Comment by antonioglass** [comment]
Hey @Fifciu ! Does something have to be configured besides just installing your version with 3DS (https://github.com/new-fantastic/vsf-payment-stripe/commits/master)? We installed it but it works just the same way as without 3DS, so no pop up with 3DS authentication appears (of course I use a 3DS test card). Thanks!

**Comment by antonioglass** [comment]
Update: we looked through the code and it looks like 3DS is triggered when ClientSecret is not null but it’s all the time null. 
![345345345](https://user-images.githubusercontent.com/71437038/116673956-af924580-a9ac-11eb-9427-9032fbf70c65.PNG)

**Comment by Fifciu** [comment]
So to make it work. I had to do a few changes in the core. 
- At first I had to use `paymentAndOrder` Magento2 endpoint instead of `order` if payment method is `stripe_payments` - it is because `paymentAndOrder` is able to give additional steps like redirect or token for additional authentication in the response. Here I also served 202 code in the response (I modified o2m.js in vsf-api for that)
- As you have additional data in the response. You have to store it somehow. As I did not have enough time in that project, I modified `order`'s core module - I added a condition inside processOrder to check if code equals 202 and there is `client_secret` in the `task.result` if so: I removed hash from order history (I will have to call order once again, after 3ds auth and without removing it would not be possible), I `setClientSecret` with `task.result.client_secret` as a value and returned `task`
- After that, `client_secret` will be updated properly and watcher in `PaymentStripe.vue` component should call `check3ds` method and handle the rest

Many things have been done there "on the asap". So there is definitely a place for improvements. Description from above is for minimal version, but for publication it would need a refactor in my opinion.

I hope you will find my tips useful

**Comment by meghasharma-24** [comment]
@Fifciu Where you added mentioned changes for comment https://github.com/develodesign/vsf-payment-stripe/issues/50#issuecomment-831832569?

**Comment by Fifciu** [comment]
@meghasharma-24 In the previous job for the client, I do not have access to this code anymore and not enough time to redevelop it. That's why I've prepared steps on how to do it so devs have a guide.

**Comment by dearste** [comment]
Any news abount this? There's a version of this module working with 3ds?

**Comment by dearste** [comment]
@meghasharma-24 have u implemented 3ds?
